### ISSUE TITLE
Handling empty bodies of wrapped functions

### ISSUE THREAD

**Issue by alexchandel** [microsoft/pyright#9613]
**Describe the bug**
The [magentic](https://github.com/jackmpcollins/magentic) package provides a few decorators that turn empty functions into advanced calls to LLMs. Here is the simplest:

```py
from magentic import prompt

@prompt('Add more "dude"ness to: {phrase}')
def dudeify(phrase: str) -> str: ...  # No function body as this is never executed

dudeify("Hello, how are you?")
# "Hey, dude! What's up? How's it going, my man?"
```

Essentially the decorator (summarized below) uses `update_wrapper` or `@wraps` to transfer the wrapped function's metadata onto a dynamically generated function that performs the call.

Now in the above example, Pyright correctly does not lint the empty `dudeify` function due to the ellipses.  However, if we provide a docstring for it, like

```py
@prompt('Add more "dude"ness to: {phrase}')
def dudeify(phrase: str) -> str: """Add more dudeness to phrase."""
```

then Pyright complains ("Function with declared return type "str" must return value on all code paths
  "None" is not assignable to "str" Pylance [reportReturnType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportReturnType)"), and if one adds an ellipse to satisfy it,

Ideally Pyright should be able to determine that this function body is thrown away / ignorable.  If not, how can we nicely annotate such a function having a docstring, indicating its body will be thrown out / never called?  Maybe a better decorator type?

**Code or Screenshots**

```py
def prompt(
    template: str,
    functions: list[Callable[..., Any]] | None = None,
    stop: list[str] | None = None,
    max_retries: int = 0,
    model: ChatModel | None = None,
) -> PromptDecorator:
    """Convert a function into an LLM prompt template.

    The `@prompt` decorator allows you to define a template for a Large Language Model (LLM) prompt as a function.
    When this function is called, the arguments are inserted into the template, then this prompt is sent to an LLM which
    generates the function output.

    Examples
    --------
    >>> @prompt("Add more dudeness to: {phrase}")
    >>> def dudeify(phrase: str) -> str: ...
    >>> # No function body as this is never executed
    >>>
    >>> dudeify("Hello, how are you?")
    "Hey, dude! What's up? How's it going, my man?"
    """

    def decorator(
        func: Callable[P, Awaitable[R]] | Callable[P, R],
    ) -> AsyncPromptFunction[P, R] | PromptFunction[P, R]:
        func_signature = inspect.signature(func)

        # ...
        # async case removed for simplicity
        # ...

        prompt_function = PromptFunction[P, R](
            name=func.__name__,
            parameters=list(func_signature.parameters.values()),
            return_type=func_signature.return_annotation,
            template=template,
            functions=functions,
            stop=stop,
            max_retries=max_retries,
            model=model,
        )
        return cast(PromptFunction[P, R], update_wrapper(prompt_function, func))

    return cast(PromptDecorator, decorator)

```

**VS Code extension or command-line**
Latest Pylance, lastest VS Code

**Comment by erictraut** [comment]
The typing spec defines a `@no_type_check` type checker directive. If you add this to the function, pyright will not report type violations within that function body. Here's a [link](https://typing.readthedocs.io/en/latest/spec/directives.html#no-type-check) to the relevant typing spec section.
